runtime/lua/vim/lsp/rpc.lua

```diff
@@ -16,102 +16,44 @@ local function format_message_with_content_length(message)
-    if key then
-      key = key:lower():gsub('%-', '_') --- @type string
-      headers[key] = value
-    else
-      log.error('invalid header line %q', line)
-      error(string.format('invalid header line %q', line))
+    local key, value = line:match('^%s*(%S+)%s*:%s*(%d+)%s*$')
+    if key and key:lower() == 'content-length' then
+      return assert(tonumber(value))
     end
   end
-  headers.content_length = tonumber(headers.content_length)
-    or error(string.format('Content-Length not found in headers. %q', header))
-  return headers
+  error('Content-Length not found in header: ' .. header)
 end
 
--- This is the start of any possible header patterns. The gsub converts it to a
--- case insensitive pattern.
-local header_start_pattern = ('content'):gsub('%w', function(c)
-  return '[' .. c .. c:upper() .. ']'
-end)
+local strbuffer = require('vim._stringbuffer')
 
---- The actual workhorse.
 local function request_parser_loop()
-  local buffer = '' -- only for header part
+  local buf = strbuffer.new()
   while true do
-    -- A message can only be complete if it has a double CRLF and also the full
-    -- payload, so first let's check for the CRLFs
-    local start, finish = buffer:find('\r\n\r\n', 1, true)
-    -- Start parsing the headers
-    if start then
-      -- This is a workaround for servers sending initial garbage before
-      -- sending headers, such as if a bash script sends stdout. It assumes
-      -- that we know all of the headers ahead of time. At this moment, the
-      -- only valid headers start with "Content-*", so that's the thing we will
-      -- be searching for.
-      -- TODO(ashkan) I'd like to remove this, but it seems permanent :(
-      local buffer_start = buffer:find(header_start_pattern)
-      if not buffer_start then
-        error(
-          string.format(
-            "Headers were expected, a different response was received. The server response was '%s'.",
-            buffer
-          )
-        )
-      end
-      local headers = parse_headers(buffer:sub(buffer_start, start - 1))
-      local content_length = headers.content_length
-      -- Use table instead of just string to buffer the message. It prevents
-      -- a ton of strings allocating.
-      -- ref. http://www.lua.org/pil/11.6.html
-      ---@type string[]
-      local body_chunks = { buffer:sub(finish + 1) }
-      local body_length = #body_chunks[1]
-      -- Keep waiting for data until we have enough.
-      while body_length < content_length do
-        ---@type string
+    local msg = buf:tostring()
+    local header_end = msg:find('\r\n\r\n', 1, true)
+    if header_end then
+      local header = buf:get(header_end + 1)
+      buf:skip(2) -- skip past header boundary
+      local content_length = get_content_length(header)
+      while strbuffer.len(buf) < content_length do
         local chunk = coroutine.yield()
-          or error('Expected more data for the body. The server may have died.') -- TODO hmm.
```

Comment:
Is this situation no longer possible? Note that `StrBuffer:put()` currently doesn't handle `nil` arg.